### DISCUSSION TITLE
Usage of "user_id" vs "sub" in JWT claims

### DISCUSSION THREAD

**Discussion by jtv8** [fastapi-users/fastapi-users#954] [General]
Quick JWT strategy related question: is there a good reason why FastAPI-Users uses "user_id" instead of "sub" for the user UUID?

From RFC 7519:

> ### [4.1.2](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.2).  "sub" (Subject) Claim
> The "sub" (subject) claim identifies the principal that is the subject of the JWT.  The claims in a JWT are normally statements about the subject.  The subject value MUST either be scoped to be locally unique in the context of the issuer or be globally unique. The processing of this claim is generally application specific.  The "sub" value is a case-sensitive string containing a StringOrURI value.  Use of this claim is OPTIONAL.

While this isn't technically invalid, it is non-standard:

> ### [4.3](https://datatracker.ietf.org/doc/html/rfc7519#section-4.3).  Private Claim Names
> A producer and consumer of a JWT MAY agree to use Claim Names that are Private Names: names that are not Registered Claim Names ([Section 4.1](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1)) or Public Claim Names ([Section 4.2](https://datatracker.ietf.org/doc/html/rfc7519#section-4.2)). Unlike Public Claim Names, Private Claim Names are subject to collision and should be used with caution.

If not, would it be advisable to switch to using "sub" in a future version? (Probably a major version increment, because I'd argue this is a compatibility-breaking change.)

**Comment by frankie567** [comment]
I fear there's no good reason 🙃 I think I misunderstood the purpose of the `sub` parameter when I first implemented this.

Open to change it to `sub` everywhere in the code base, but keep the `user_id` claim in the JWT in case some people are decoding the JWT and relying on it.
